lib/jackhammer/middleware_collection.rb

```diff
@@ -0,0 +1,35 @@
+module Jackhammer
+  class MiddlewareCollection
+    def initialize
+      @entries = []
+    end
+
+    def use(klass, *args, &block)
+      @entries << Entry.new(klass: klass, args: args, block: block)
+    end
+
+    def call(*args)
+      return yield *args if @entries.empty?
+
+      middlewares = @entries.map(&:instantiate)
+      traverse = proc do |*arguments|
+        if middlewares.empty?
+          yield *arguments
```

Comment:
At first glance I didn't like the use of `yield` in `.call` because it makes the code a little harder to reason about for me personally, but the more I see it in action the more I like it 👍  I'm curious what is the motivation to use this technique over calling a final proc/function injected just prior to traversing?

Reply:
Is this roughly what you suggest as the alternative?

```ruby
def call(*args, &block)
  middlewares = @entries.map(&:instantiate) + [block]
  traverse = proc do |*arguments|
    middlewares.shift.call(*arguments, &traverse) unless middlewares.empty?
  end
  traverse.call(*args)
end
```

If so, then I guess my problem with that was that we lose the result of the block that is passed to call. Besides that, I think it's indeed more readable.

Personally, I'd prefer to name the block passed in to call and call it directly (`block.call` instead of `yield`) because I always forget whether you can use yield in a proc and have to think a bit on whether this will refer to the block passed to the proc or to the one passed to the method that defines it, but `rubocop` does not allow that.

Reply:
Actually, I was wrong - you don't lose the result because you wouldn't yield from the block passed in to `call` so that `middlewares.empty?` condition is never `true`.

Reply:
I was thinking like how `Rack::Builder` works where you have the same interface with `use` but then the client of the library implements the last chain in the middleware to a separate method that generates the chain
```ruby
builder = Rack::Builder.new
builder.user Middleware.new arg1
chain = builder.run lambda { |env| [200, { "Content-Type" => "text/plain" }, ["OK"]] }
chain.call(message)
```
However, the more I think about it the more I think Jackhammer doesn't need a similar `run` interface because the final handler isn't something to customize by users of this library..

Reply:
Ah, yes - we don't need that here, since the last step in the chain is fixed and is the code the library is supposed to execute.